### PR TITLE
Json Frontend Tests

### PR DESCRIPTION
**Pull request by cwild-UoS**
Draft PR for now until all tests are written

### CODE REVIEW

**Review by bobturneruk — APPROVED** [review]
These are positive changes. There are a few questions in there. Main things:

- Looks like some important code remains uncovered.
- Some "private" methods are tested directly, some aren't.
- Some "public" methods aren't tested directly as units.

It would be good to capture the reasoning for this, here, in issues, or comments in the test code, to aid future developers who might break the tests when altering the code, or seek to extend the test set in future.

File: tests/json_front_tests/test_json_class.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+    def test_set_metas(self):
+        self.json_class.set_variables(self.input_dict_list, self.output_dict_list, self.meta_dict_list)
+        ctf_meta = [Meta("test_meta", float, populate_example)]
+        self.assertEqual(ctf_meta[0].name, self.json_class.metas[0].name)
+        self.assertEqual(ctf_meta[0].datatype, self.json_class.metas[0].datatype)
+
+    def test_argparse(self):
+        args = self.json_class.get_args(["--data_path=data.csv", "--dag_path=dag.dot", "--json_path=tests.json"])
+        self.assertTrue(args.data_path)
```
[the remark this answers is not in the document]
Comment (cwild-UoS) [reply tests/json_front_tests/test_json_class.py:65]: Seems better to me as well, cheers for suggestion :)

File: tests/json_front_tests/test_json_class.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        self.assertTrue(args.dag_path)
+        self.assertTrue(args.json_path)
+
+    def test_setup_modelling_scenario(self):
+        self.json_class.setup()
```
[the remark this answers is not in the document]
Comment (cwild-UoS) [reply tests/json_front_tests/test_json_class.py:70]: It would be cleaner code wise to place it in setUp. It's quite computationally expensive to call though and on my laptop roughly doubles the test execute time to call with each test from ~1.6 s to ~3s . Not sure what is better in this tradeoff?

[the remark this answers is not in the document]
Comment (bobturneruk) [reply tests/json_front_tests/test_json_class.py:70]: Is `setUp` only called once, or for each test?

[the remark this answers is not in the document]
Comment (cwild-UoS) [reply tests/json_front_tests/test_json_class.py:70]: For each test currently

[the remark this answers is not in the document]
Comment (cwild-UoS) [reply tests/json_front_tests/test_json_class.py:70]: Would things like the path file names be better off as global variables? Rather than being redefined for each test

[the remark this answers is not in the document]
Comment (bobturneruk) [reply tests/json_front_tests/test_json_class.py:70]: I think best practise is to use `setUp`, but I mostly use `pytest` where this would probably be a fixture.

File: tests/json_front_tests/test_json_class.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+    def tearDown(self) -> None:
+        remove_temp_dir_if_existent()
+
+
+def populate_example(*args, **kwargs):
+    pass
+
+
+def setup_json_file(json_path):
```
[the remark this answers is not in the document]
Comment (cwild-UoS) [reply tests/json_front_tests/test_json_class.py:135]: Ah okay, good to know! Personally I prefer having it in the code as it is a relatively simple setup and it was useful for me to reference to when looking at the tests. 
Is this more of a preference rather than a 'right/wrong' trade-off then?

[the remark this answers is not in the document]
Comment (bobturneruk) [reply tests/json_front_tests/test_json_class.py:135]: I think file size / computational cost would heavily influence decision if they were issues.

[the remark this answers is not in the document]
Comment (cwild-UoS) [reply tests/json_front_tests/test_json_class.py:135]: Makes sense thanks :). For now it is a very small json file with only a few lines in it.

The current setup creates and deletes this file for each test though. So I'll follow your suggestion as the more scalable and cleaner format! 

Cheers bob

File: causal_testing/json_front/json_class.py [hunk 4]
```diff
@@ -89,54 +88,58 @@ def setup(self):
...
-        logger.info("{%d}/{%d} failed", failures, executed_tests)
+    def _execute_tests(self, concrete_tests, estimators, test, f_flag):
+        failures = 0
+        for concrete_test in concrete_tests:
+            failed = self._execute_test_case(concrete_test, estimators[test["estimator"]], f_flag)
+            if failed:
+                failures += 1
+        return failures
 
     def _json_parse(self):
```
[the remark this answers is not in the document]
Comment (cwild-UoS) [reply causal_testing/json_front/json_class.py:138]: After reading your comments on testing a mix of both public/private methods and how hard to follow that got. I've moved to just testing the public methods as 'units'. I couldn't find any other private method testing in the project so I think it will be nice to follow the rest of the projects testing as well.

[the remark this answers is not in the document]
Comment (cwild-UoS) [reply causal_testing/json_front/json_class.py:138]: So this specific code is now covered in the testing of `generate_tests`
